wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/validator/DevfileSchemaValidator.java

```diff
@@ -11,48 +11,66 @@
  */
 package org.eclipse.che.api.devfile.server.validator;
 
+import static java.lang.String.format;
+
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
-import com.github.fge.jsonschema.core.exceptions.ProcessingException;
-import com.github.fge.jsonschema.core.report.ProcessingReport;
-import com.github.fge.jsonschema.main.JsonSchemaFactory;
-import com.github.fge.jsonschema.main.JsonValidator;
 import java.io.IOException;
+import java.io.StringReader;
+import java.util.ArrayList;
+import java.util.List;
 import javax.inject.Inject;
 import javax.inject.Singleton;
+import javax.json.JsonReader;
 import org.eclipse.che.api.devfile.server.exception.DevfileFormatException;
 import org.eclipse.che.api.devfile.server.schema.DevfileSchemaProvider;
+import org.leadpony.justify.api.JsonSchema;
+import org.leadpony.justify.api.JsonValidationService;
+import org.leadpony.justify.api.Problem;
+import org.leadpony.justify.api.ProblemHandler;
 
 /** Validates YAML devfile content against given JSON schema. */
 @Singleton
 public class DevfileSchemaValidator {
 
-  private JsonValidator validator;
+  private static final JsonValidationService service = JsonValidationService.newInstance();
   private ObjectMapper yamlReader;
-  private DevfileSchemaProvider schemaProvider;
-  private ErrorMessageComposer messageComposer;
+  private ObjectMapper jsonWriter;
+  private JsonSchema schema;
+  private ErrorMessageComposer errorMessageComposer;
 
   @Inject
   public DevfileSchemaValidator(DevfileSchemaProvider schemaProvider) {
-    this.schemaProvider = schemaProvider;
-    this.validator = JsonSchemaFactory.byDefault().getValidator();
     this.yamlReader = new ObjectMapper(new YAMLFactory());
-    this.messageComposer = new ErrorMessageComposer();
+    this.jsonWriter = new ObjectMapper();
+    this.errorMessageComposer = new ErrorMessageComposer();
+    try {
+      this.schema = service.readSchema(schemaProvider.getAsReader());
+    } catch (IOException e) {
+      throw new RuntimeException("Unable to read devfile json schema for validation.", e);
+    }
   }
 
   public JsonNode validateBySchema(String yamlContent) throws DevfileFormatException {
-    ProcessingReport report;
-    JsonNode data;
+    JsonNode contentNode;
     try {
-      data = yamlReader.readTree(yamlContent);
-      report = validator.validate(schemaProvider.getJsoneNode(), data);
-    } catch (IOException | ProcessingException e) {
+      contentNode = yamlReader.readTree(yamlContent);
+      List<Problem> validationErrors = new ArrayList<>();
+      ProblemHandler handler = ProblemHandler.collectingTo(validationErrors);
+      try (JsonReader reader =
+          service.createReader(
+              new StringReader(jsonWriter.writeValueAsString(contentNode)), schema, handler)) {
```

Comment:
Do you really need to parse the content and then convert it back to YAML? I assume it may be useful if YAML is not a valid, but please check maybe justify is able to handle it by itself. It code would be much simpler.
```java
class DevfileManager {...
  public DevfileImpl parse(String devfileContent) throws DevfileFormatException {
    schemaValidator.validateBySchema(devfileContent);

    DevfileImpl devfile;
    try {
      devfile = objectMapper.readValue(devfileContent, DevfileImpl.class);
    } catch (IOException e) {
      throw new DevfileFormatException(e.getMessage());
    }

    integrityValidator.validateDevfile(devfile);

    return devfile;
  }
}
```
```java
public class DevfileSchemaValidator {

  private static final JsonValidationService service = JsonValidationService.newInstance();
  private JsonSchema schema;
  private ErrorMessageComposer errorMessageComposer;

  @Inject
  public DevfileSchemaValidator(DevfileSchemaProvider schemaProvider) {
    this.errorMessageComposer = new ErrorMessageComposer();
    try {
      this.schema = service.readSchema(schemaProvider.getAsReader());
    } catch (IOException e) {
      throw new RuntimeException("Unable to read devfile json schema for validation.", e);
    }
  }

  public void validateBySchema(String yamlContent) throws DevfileFormatException {
    try {
      List<Problem> validationErrors = new ArrayList<>();
      ProblemHandler handler = ProblemHandler.collectingTo(validationErrors);
      try (JsonReader reader = service.createReader(new StringReader(yamlContent), schema, handler)) {
        reader.read();
      }

      if (!validationErrors.isEmpty()) {
        String error = errorMessageComposer.extractMessages(validationErrors, new StringBuilder());
        throw new DevfileFormatException(
            format("Devfile schema validation failed. Error: %s", error));
      }

    } catch (JSOnException e) {
      throw new DevfileFormatException("Unable to validate Devfile. Error: " + e.getMessage());
    }
  }
}
```

Reply:
There is a few reasons preventing from making it in such way:
0)  None of the validators can work with yaml directly. We need to feed them with json :-)
1) Avoid double yaml-json (validator) and yaml-> json -> pojo (manager) conversion.

So as for now,  we first convert yaml -> JsonNode, then pass this node as a string to validator (not converting back to yaml), and pass the same node to Manager to create POJO from it.